pkg/ingester/ingester.go

```diff
@@ -714,16 +725,54 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, pushReq *push.Request) (
 	)
 
 	// Walk the samples, appending them to the users database
-	app := db.Appender(ctx).(extendedAppender)
-	level.Debug(spanlog).Log("event", "got appender", "numSeries", len(req.Timeseries))
+	var persistentApp, ephemeralApp extendedAppender
 
-	var activeSeries *activeseries.ActiveSeries
-	if i.cfg.ActiveSeriesMetricsEnabled {
-		activeSeries = db.activeSeries
+	rollback := func() {
+		if persistentApp != nil {
+			if err := persistentApp.Rollback(); err != nil {
+				level.Warn(i.logger).Log("msg", "failed to rollback on error", "user", userID, "err", err)
+			}
+		}
+		if ephemeralApp != nil {
+			if err := ephemeralApp.Rollback(); err != nil {
+				level.Warn(i.logger).Log("msg", "failed to rollback ephemeral appender on error", "user", userID, "err", err)
+			}
+		}
 	}
-	err = i.pushSamplesToAppender(userID, req.Timeseries, app, startAppend, &stats, updateFirstPartial, activeSeries, i.limits.OutOfOrderTimeWindow(userID), minAppendTimeAvailable, minAppendTime)
-	if err != nil {
-		return nil, err
+
+	if len(req.Timeseries) > 0 {
+		persistentApp = db.Appender(ctx).(extendedAppender)
+
+		level.Debug(spanlog).Log("event", "got appender", "series", len(req.Timeseries))
+
+		var activeSeries *activeseries.ActiveSeries
+		if i.cfg.ActiveSeriesMetricsEnabled {
+			activeSeries = db.activeSeries
+		}
+		err = i.pushSamplesToAppender(userID, req.Timeseries, persistentApp, startAppend, &stats, updateFirstPartial, activeSeries, i.limits.OutOfOrderTimeWindow(userID), minAppendTimeAvailable, minAppendTime, true)
+		if err != nil {
+			rollback()
+			return nil, err
+		}
+	}
+
+	if len(req.EphemeralTimeseries) > 0 {
+		a, err := db.EphemeralAppender(ctx)
+		if err != nil {
+			// TODO: handle error caused by limit (ephemeral storage disabled), and report it via firstPartialErr instead.
+			rollback()
+			return nil, err
+		}
+
+		ephemeralApp = a.(extendedAppender)
+
+		level.Debug(spanlog).Log("event", "got appender for ephemeral series", "ephemeralSeries", len(req.EphemeralTimeseries))
+
+		err = i.pushSamplesToAppender(userID, req.EphemeralTimeseries, ephemeralApp, startAppend, &ephemeralStats, updateFirstPartial, nil, 0, minAppendTimeAvailable, minAppendTime, false)
```

Comment:
in the future we might want to consider accept out-of-order samples here as well, but yeah let's ignore that for now